stores/sql/metrics.go

```diff
@@ -3,79 +3,221 @@ package sql
+const (
+	contractMetricGranularity = 5 * time.Minute
+)
+
+type (
+	ContractMetricsQueryOpts struct {
+		api.ContractMetricsQueryOpts
+		IndexHint string
 	}
-	defer rows.Close()
+)
 
-	for rows.Next() {
-		var cpm api.ContractPruneMetric
+func ContractMetrics(ctx context.Context, tx sql.Tx, start time.Time, n uint64, interval time.Duration, opts ContractMetricsQueryOpts) (metrics []api.ContractMetric, err error) {
+	// define a helper function to scan a contract metric from a row.
+	scanContractMetric := func(rows *sql.LoggedRows, aggregate bool) (cm api.ContractMetric, err error) {
 		var placeHolder int64
 		var placeHolderTime time.Time
+		var placeHolderHK PublicKey
+		var placeHolderFCID FileContractID
+		var placeHolderRevisionNumber Unsigned64
+
 		var timestamp UnixTimeMS
-		if err := rows.Scan(&placeHolder, &placeHolderTime, &timestamp, (*FileContractID)(&cpm.ContractID), (*PublicKey)(&cpm.HostKey), &cpm.HostVersion, (*Unsigned64)(&cpm.Pruned), (*Unsigned64)(&cpm.Remaining), &cpm.Duration); err != nil {
-			return nil, fmt.Errorf("failed to scan contract prune metric: %w", err)
+		err = rows.Scan(
+			&placeHolder,
+			&placeHolderTime,
+			&timestamp,
+			&placeHolderFCID,
+			&placeHolderHK,
+			(*Unsigned64)(&cm.RemainingCollateral.Lo), (*Unsigned64)(&cm.RemainingCollateral.Hi),
+			(*Unsigned64)(&cm.RemainingFunds.Lo), (*Unsigned64)(&cm.RemainingFunds.Hi),
+			&placeHolderRevisionNumber,
+			(*Unsigned64)(&cm.UploadSpending.Lo), (*Unsigned64)(&cm.UploadSpending.Hi),
+			(*Unsigned64)(&cm.DownloadSpending.Lo), (*Unsigned64)(&cm.DownloadSpending.Hi),
+			(*Unsigned64)(&cm.FundAccountSpending.Lo), (*Unsigned64)(&cm.FundAccountSpending.Hi),
+			(*Unsigned64)(&cm.DeleteSpending.Lo), (*Unsigned64)(&cm.DeleteSpending.Hi),
+			(*Unsigned64)(&cm.ListSpending.Lo), (*Unsigned64)(&cm.ListSpending.Hi),
+		)
+		if err != nil {
 		}
-		cpm.Timestamp = api.TimeRFC3339(normaliseTimestamp(start, interval, timestamp))
-		metrics = append(metrics, cpm)
+
+		cm.Timestamp = api.TimeRFC3339(normaliseTimestamp(start, interval, timestamp))
+		if !aggregate {
+			cm.ContractID = types.FileContractID(placeHolderFCID)
+			cm.HostKey = types.PublicKey(placeHolderHK)
+			cm.RevisionNumber = uint64(placeHolderRevisionNumber)
+		}
+		return
 	}
 
-	return metrics, nil
+	// if a host filter is set, query periods
+	if opts.ContractID != (types.FileContractID{}) || opts.HostKey != (types.PublicKey{}) {
+		err = queryPeriods(ctx, tx, start, n, interval, opts.ContractMetricsQueryOpts, func(rows *sql.LoggedRows) error {
+			cm, err := scanContractMetric(rows, false)
+			if err != nil {
+				return fmt.Errorf("failed to scan contract metrics: %w", err)
+			}
+			metrics = append(metrics, cm)
+			return nil
+		})
+		return
+	}
+
+	// otherwise we return the aggregated metrics for each period
+	currentPeriod := int64(math.MinInt64)
+	err = queryAggregatedPeriods(ctx, tx, start, n, interval, opts.IndexHint, func(period int64, rows *sql.LoggedRows) error {
```

Comment:
`queryAggregatedPeriods` is only called in one place so it should not need the callback at all right? 

That way both `queryPeriods` and `queryAggregatedPeriods` could return a slice already instead of appending here. Assuming you tweak `queryPeriods` with the generic param.